plugins/storage/fileshare/storagefsvm/src/main/java/org/apache/cloudstack/storage/fileshare/lifecycle/StorageFsVmFileShareLifeCycle.java

```diff
@@ -0,0 +1,329 @@
+                customParameterMap, null, null, null, null,
+                true, UserVmManager.STORAGEFSVM, null);
+        return vm;
+    }
+
+    @Override
+    public void checkPrerequisites(DataCenter zone, Long serviceOfferingId) {
+        ServiceOfferingVO serviceOffering = serviceOfferingDao.findById(serviceOfferingId);
+        if (serviceOffering.getCpu() < STORAGEFSVM_MIN_CPU_COUNT.valueIn(zone.getId())) {
+            throw new InvalidParameterValueException("Service offering's number of cpu should be greater than or equal to " + STORAGEFSVM_MIN_CPU_COUNT.key());
+        }
+        if (serviceOffering.getRamSize() < STORAGEFSVM_MIN_RAM_SIZE.valueIn(zone.getId())) {
+            throw new InvalidParameterValueException("Service offering's ram size should be greater than or equal to " + STORAGEFSVM_MIN_RAM_SIZE.key());
+        }
+        if (serviceOffering.isOfferHA() == false) {
+            throw new InvalidParameterValueException("Service offering's should be HA enabled");
+        }
+
+        Hypervisor.HypervisorType availableHypervisor = resourceMgr.getAvailableHypervisor(zone.getId());
+        VMTemplateVO template = templateDao.findSystemVMReadyTemplate(zone.getId(), availableHypervisor);
+        if (template == null) {
+            throw new CloudRuntimeException(String.format("Unable to find the system templates or it was not downloaded in %s.", zone.toString()));
+        }
+    }
+
+    @Override
+    public Pair<Long, Long> commitFileShare(FileShare fileShare, Long networkId, Long diskOfferingId, Long size, Long minIops, Long maxIops) throws ResourceUnavailableException, InsufficientCapacityException, ResourceAllocationException {
+        Account owner = accountMgr.getActiveAccountById(fileShare.getAccountId());
+        UserVm vm = createFileShareVM(fileShare.getDataCenterId(), owner, List.of(networkId), fileShare.getName(), fileShare.getServiceOfferingId(), diskOfferingId, fileShare.getFsType(), size, minIops, maxIops);
+
+        List<VolumeVO> volumes = volumeDao.findByInstanceAndType(vm.getId(), Volume.Type.DATADISK);
+        return new Pair<>(volumes.get(0).getId(), vm.getId());
+    }
+
+    @Override
+    public void startFileShare(FileShare fileShare) throws OperationTimedoutException, ResourceUnavailableException, InsufficientCapacityException {
+        UserVmVO vm = userVmDao.findById(fileShare.getVmId());
+        userVmManager.startVirtualMachine(vm);
+    }
+
+    @Override
+    public boolean stopFileShare(FileShare fileShare, Boolean forced) {
+        userVmManager.stopVirtualMachine(fileShare.getVmId(), false);
+        return true;
+    }
+
+    private void expungeVm(Long vmId) {
+        UserVmVO userVM = userVmDao.findById(vmId);
+        if (userVM == null) {
+            return;
+        }
+        try {
+            UserVm vm = userVmService.destroyVm(userVM.getId(), true);
+            if (!userVmManager.expunge(userVM)) {
```

Comment:
the previous line will expunge the vm, right ?

```
UserVm vm = userVmService.destroyVm(userVM.getId(), true);
```

Reply:
destroyVM doesn't really expunge the VM. expunge() has to be called to expunge it completely. see https://github.com/apache/cloudstack/blob/main/plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterScaleWorker.java#L327